index.js

```diff
@@ -28,16 +28,31 @@ function fastifyCors (fastify, opts, next) {
     optionsSuccessStatus,
     preflight,
     hideOptionsRoute,
-    strictPreflight
+    strictPreflight,
+    preflightContinue
   } = Object.assign({}, defaultOptions, opts)
 
   const resolveOriginOption = typeof origin === 'function' ? resolveOriginWrapper : (_, cb) => cb(null, origin)
 
-  fastify.decorateRequest('corsPreflightEnabled', undefined)
+  fastify.decorateRequest('corsPreflightEnabled', false)
   fastify.addHook('onRequest', onRequest)
 
   if (preflight === true) {
-    fastify.options('*', { schema: { hide: hideOptionsRoute } }, preflightHandler)
+    // The preflight reply must occur in the hook. This allows fastify-cors to reply to
+    // preflight requests BEFORE possible authentication plugins. If the preflight reply
+    // occurred in the this handler, other plugins may deny the request since the browser will
+    // remove most headers (such as the Authentication header).
+    //
+    // This route simply enables fastify to accept preflight requests.
+    fastify.options('*', { schema: { hide: hideOptionsRoute } }, (req, reply) => {
+      if (!req.corsPreflightEnabled) {
+        // Do not handle preflight requests if the origin option disabled CORS
+        reply.code(404).type('text/plain').send('Not Found')
```

Comment:
This is a new change: reply 404 if preflight was disabled but the options route still gets called. This can happen if `preflightContinue` is true but the `origin` option resolves to false.

Reply:
I think there is a `reply.notFound()` helper to send it to the not found handler.

Reply:
:clap: nice! done :boom: